v3_ex10.py

amymhaddad opened the pull request.

amymhaddad committed:
v3_ex10.py

amymhaddad committed:
Update exercise10.py

amymhaddad commented on exercise10.py:
```diff
@@ -0,0 +1,47 @@
+    return subtotal_item1 + subtotal_item2 + subtotal_item3
+
+
+def convert_tax_to_decimal(TAX_RATE):
```
@paulghaddad   I could call convert_tax_to_decimal and add it to my parameter, instead of re-calculating it below, right?

paulghaddad replied:
Not sure what you mean. However, remember `TAX_RATE` on line 30 is the _parameter_, not the specific argument you pass in. It is the general name of the input for the function. Thus, it should be lowercase here. But when you _call_ the function, you can pass in any named variable as the `argument`. Which is what you do in the `main` file:

https://github.com/amymhaddad/version_2_exercises_for_programmers/pull/35/files#diff-98e5bbb0f3420abac5ca67b81848262aR17

or in the test:

https://github.com/amymhaddad/version_2_exercises_for_programmers/pull/35/files#diff-322cd8d0a431861e5483d2fd42dd5905R39

paulghaddad commented on main_exercise10.py:
```diff
@@ -0,0 +1,44 @@
+    calculate_grand_total,
+)
+
+TAX_RATE = 5.5
```
Excellent use of a constant

paulghaddad commented on main_exercise10.py:
```diff
@@ -0,0 +1,44 @@
+    """Create a self-checkout system that calculates the grand total for three items"""
+
+    price_item1, quantity_item1 = get_price_quantity()
+    print("\n")
```
I think in Python3, there's a shortcut. You can just use `print()` to print a new line.

paulghaddad commented on main_exercise10.py:
```diff
@@ -0,0 +1,44 @@
+    subtotal_item2 = get_subtotal(price_item2, quantity_item2)
+    subtotal_item3 = get_subtotal(price_item3, quantity_item3)
+
+    sum_of_subtotals = calculate_sum_of_subtotals(subtotal_item1, subtotal_item2, subtotal_item3)
```
A nitpick: probably no need to use "calculate" in your function name. `sum_of_subtotals` or `subtotal_sum` is sufficient.

paulghaddad commented on main_exercise10.py:
```diff
@@ -0,0 +1,44 @@
+    subtotal_item3 = get_subtotal(price_item3, quantity_item3)
+
+    sum_of_subtotals = calculate_sum_of_subtotals(subtotal_item1, subtotal_item2, subtotal_item3)
+    total_tax_for_items = calculate_total_tax_for_items(TAX_RATE, sum_of_subtotals)
```
Excellent interface (what the function takes in as input) for `calculate_total_tax_for_items`. That's exactly what I'd do. However, one nit-pick is to reverse the parameter order as the subtotal is a bit more important imo that the tax rate. The more important parameters should be listed first.

paulghaddad commented on main_exercise10.py:
```diff
@@ -0,0 +1,44 @@
+
+    sum_of_subtotals = calculate_sum_of_subtotals(subtotal_item1, subtotal_item2, subtotal_item3)
+    total_tax_for_items = calculate_total_tax_for_items(TAX_RATE, sum_of_subtotals)
+    grand_total = calculate_grand_total(total_tax_for_items, sum_of_subtotals)
```
Similar comment as above: `grand_total` is sufficient for the name. Also, flip the order of the parameters by importance.

paulghaddad commented on main_exercise10.py:
```diff
@@ -0,0 +1,44 @@
+    total_tax_for_items = calculate_total_tax_for_items(TAX_RATE, sum_of_subtotals)
+    grand_total = calculate_grand_total(total_tax_for_items, sum_of_subtotals)
+
+    formatted_output = f"""
```
👏

paulghaddad commented on test_exercise10.py:
```diff
@@ -0,0 +1,72 @@
+)
+
+class SelfCheckoutTest(unittest.TestCase):
+    """Test to see if self-checkout system calculates grand total for three items"""
```
Nice comments throughout.

paulghaddad commented on test_exercise10.py:
```diff
@@ -0,0 +1,72 @@
+    def test_subtotal(self):
+        """Test to calculate subtotal"""
+
+        price = 5.0
```
Fantastic organization. I can see you're using the "Four Phases of Testing"! 💯

paulghaddad commented on test_exercise10.py:
```diff
@@ -0,0 +1,72 @@
+    calculate_grand_total,
+)
+
+class SelfCheckoutTest(unittest.TestCase):
```
All the tests here are superb. Excellent work.

paulghaddad commented on exercise10.py:
```diff
@@ -0,0 +1,47 @@
+    return price * quantity
+
+
+def calculate_sum_of_subtotals(subtotal_item1, subtotal_item2, subtotal_item3):
```
I'd recommend allowing this function to take in any number of arguments by using the splat operator: `*subtotals`. Then it's more flexible.

paulghaddad commented on exercise10.py:
```diff
@@ -0,0 +1,47 @@
+
+@author: amyhaddad
+"""
+
```
The functions in this file are named relatively well, but even more so, are super short and succinct. That's what you're striving for.

paulghaddad reviewed (changes requested):
Overall excellent work. I have several suggestions to tighten up this problem, but overall, your functions were excellent, the `main` file was well organized and easy to read top to bottom, and the tests were flawless.

amymhaddad committed:
Update exercise10.py

amymhaddad commented on exercise10.py:
```diff
@@ -0,0 +1,47 @@
+    return tax_rate / 100
+
+
+def total_tax_for_multiple_items(tax_rate_as_decimal, subtotals_sum):
```
@paulghaddad This is the function I was referring to in my previous comment. Originally I had this:

`def calculate_total_tax_for_items(TAX_RATE, sum_of_subtotals):
    """Calculate the total tax: tax rate as decimal * sum of the subtotals"""

    tax_rate_as_decimal = convert_tax_to_decimal(TAX_RATE)

    return tax_rate_as_decimal * sum_of_subtotals`

Notice that I called the `convert_tax_to_decimal` function within the function. However, this time around I took it out and passed in the function as a variable in the parameter. Which method is preferred?

amymhaddad committed:
Update main_exercise10.py

amymhaddad commented:
@paulghaddad Great comments! I updated many of the variable names and function names. And I used * to pass in any number of arguments in the parameter. I also updated two functions: `sum_of_subtotals` and `total_tax_for_multiple_items`.
